File: clients/csharp/BaseAI.cs
using System;
using System.Runtime.InteropServices;

/// <summary>
/// This class implements most the code an AI would need to interface with the lower-level game code.
/// AIs should extend this class to get a lot of builer-plate code out of the way. The provided AI class does just that.
/// </summary>
public abstract class BaseAI
{
  public static Player[] players;
  public static Mappable[] mappables;
  public static Tile[] tiles;
  public static Trap[] traps;
  public static Thief[] thiefs;
  public static ThiefType[] thiefTypes;
  public static TrapType[] trapTypes;

  IntPtr connection;
  public static int iteration;
  bool initialized;

  public BaseAI(IntPtr c)
  {
    connection = c;
  }

  /// <summary>
  /// Make this your username, which should be provided.
  /// </summary>
  /// <returns>Returns the username of the player.</returns>
  public abstract String username();

  /// <summary>
  /// Make this your password, which should be provided.
  /// </summary>
  /// <returns>Returns the password of the player.</returns>
  public abstract String password();

  /// <summary>
  /// This is run once on turn one before run().
  /// </summary>
  public abstract void init();

  /// <summary>
  /// This is run every turn.
  /// </summary>
  /// <returns>
  /// Return true to end turn, false to resynchronize with the 
  /// server and run again.
  /// </returns>
  public abstract bool run();

  /// <summary>
  /// This is run once after your last turn.
  /// </summary>
  public abstract void end();

  /// <summary>
  /// Synchronizes with the server, then calls run().
  /// </summary>
  /// <returns>
  /// Return true to end turn, false to resynchronize with the 
  /// server and run again.
  /// </returns>
  public bool startTurn()
  {
    int count = 0;
    iteration++;

    count = Client.getPlayerCount(connection);
    players = new Player[count];
    for(int i = 0; i < count; i++)
      players[i] = new Player(Client.getPlayer(connection, i));

    count = Client.getMappableCount(connection);
    mappables = new Mappable[count];
    for(int i = 0; i < count; i++)
      mappables[i] = new Mappable(Client.getMappable(connection, i));

    count = Client.getTileCount(connection);
    tiles = new Tile[count];
    for(int i = 0; i < count; i++)
      tiles[i] = new Tile(Client.getTile(connection, i));

    count = Client.getTrapCount(connection);
    traps = new Trap[count];
    for(int i = 0; i < count; i++)
      traps[i] = new Trap(Client.getTrap(connection, i));

    count = Client.getThiefCount(connection);
    thiefs = new Thief[count];
    for(int i = 0; i < count; i++)
      thiefs[i] = new Thief(Client.getThief(connection, i));

    count = Client.getThiefTypeCount(connection);
    thiefTypes = new ThiefType[count];
    for(int i = 0; i < count; i++)
      thiefTypes[i] = new ThiefType(Client.getThiefType(connection, i));

    count = Client.getTrapTypeCount(connection);
    trapTypes = new TrapType[count];
    for(int i = 0; i < count; i++)
      trapTypes[i] = new TrapType(Client.getTrapType(connection, i));

    if(!initialized)
    {
      initialized = true;
      init();
    }

    return run();
  }

  /// <summary>
  /// The width of the total map.
  /// </summary>
  /// <returns>Returns the width of the total map.</returns>
  public int mapWidth()
  {
    int value = Client.getMapWidth(connection);
    return value;
  }

  /// <summary>
  /// The height of the total map.
  /// </summary>
  /// <returns>Returns the height of the total map.</returns>
  public int mapHeight()
  {
    int value = Client.getMapHeight(connection);
    return value;
  }

  /// <summary>
  /// The current turn number.
  /// </summary>
  /// <returns>Returns the current turn number.</returns>
  public int turnNumber()
  {
    int value = Client.getTurnNumber(connection);
    return value;
  }

  /// <summary>
  /// The current turn number for this round.
  /// </summary>
  /// <returns>Returns the current turn number for this round.</returns>
  public int roundTurnNumber()
  {
    int value = Client.getRoundTurnNumber(connection);
    return value;
  }

  /// <summary>
  /// The maximum number of Thieves allowed per player.
  /// </summary>
  /// <returns>Returns the maximum number of Thieves allowed per player.</returns>
  public int maxThieves()
  {
    int value = Client.getMaxThieves(connection);
    return value;
  }

  /// <summary>
  /// The maximum number of Traps allowed per player.
  /// </summary>
  /// <returns>Returns the maximum number of Traps allowed per player.</returns>
  public int maxTraps()
  {
    int value = Client.getMaxTraps(connection);
    return value;
  }

  /// <summary>
  /// The id of the current player.
  /// </summary>
  /// <returns>Returns the id of the current player.</returns>
  public int playerID()
  {
    int value = Client.getPlayerID(connection);
    return value;
  }

  /// <summary>
  /// What number game this is for the server.
  /// </summary>
  /// <returns>Returns what number game this is for the server.</returns>
  public int gameNumber()
  {
    int value = Client.getGameNumber(connection);
    return value;
  }

  /// <summary>
  /// What round of the game this is.
  /// </summary>
  /// <returns>Returns what round of the game this is.</returns>
  public int roundNumber()
  {
    int value = Client.getRoundNumber(connection);
    return value;
  }

  /// <summary>
  /// The scarabs given to a player to purchase traps per round.
  /// </summary>
  /// <returns>Returns the scarabs given to a player to purchase traps per round.</returns>
  public int scarabsForTraps()
  {
    int value = Client.getScarabsForTraps(connection);
    return value;
  }

  /// <summary>
  /// The scarabs given to a player to purchase thieves per round.
  /// </summary>
  /// <returns>Returns the scarabs given to a player to purchase thieves per round.</returns>
  public int scarabsForThieves()
  {
    int value = Client.getScarabsForThieves(connection);
    return value;
  }

  /// <summary>
  /// The maximum number of thieves per tile.
  /// </summary>
  /// <returns>Returns the maximum number of thieves per tile.</returns>
  public int maxStack()
  {
    int value = Client.getMaxStack(connection);
    return value;
  }

  /// <summary>
  /// The number of won rounds required to win.
  /// </summary>
  /// <returns>Returns the number of won rounds required to win.</returns>
  public int roundsToWin()
  {
    int value = Client.getRoundsToWin(connection);
    return value;
  }
}

File: clients/java/BaseAI.java
import com.sun.jna.Pointer;

/// \brief A basic AI interface.

///This class implements most the code an AI would need to interface with the lower-level game code.
///AIs should extend this class to get a lot of builer-plate code out of the way
///The provided AI class does just that.
public abstract class BaseAI
{
  static Player[] players;
  static Mappable[] mappables;
  static Tile[] tiles;
  static Trap[] traps;
  static Thief[] thiefs;
  static ThiefType[] thiefTypes;
  static TrapType[] trapTypes;
  Pointer connection;
  static int iteration;
  boolean initialized;

  public BaseAI(Pointer c)
  {
    connection = c;
  }
    
  ///
  ///Make this your username, which should be provided.
  public abstract String username();
  ///
  ///Make this your password, which should be provided.
  public abstract String password();
  ///
  ///This is run on turn 1 before run
  public abstract void init();
  ///
  ///This is run every turn . Return true to end the turn, return false
  ///to request a status update from the server and then immediately rerun this function with the
  ///latest game status.
  public abstract boolean run();

  ///
  ///This is run on after your last turn.
  public abstract void end();


  public boolean startTurn()
  {
    iteration++;
    int count = 0;
    count = Client.INSTANCE.getPlayerCount(connection);
    players = new Player[count];
    for(int i = 0; i < count; i++)
    {
      players[i] = new Player(Client.INSTANCE.getPlayer(connection, i));
    }
    count = Client.INSTANCE.getMappableCount(connection);
    mappables = new Mappable[count];
    for(int i = 0; i < count; i++)
    {
      mappables[i] = new Mappable(Client.INSTANCE.getMappable(connection, i));
    }
    count = Client.INSTANCE.getTileCount(connection);
    tiles = new Tile[count];
    for(int i = 0; i < count; i++)
    {
      tiles[i] = new Tile(Client.INSTANCE.getTile(connection, i));
    }
    count = Client.INSTANCE.getTrapCount(connection);
    traps = new Trap[count];
    for(int i = 0; i < count; i++)
    {
      traps[i] = new Trap(Client.INSTANCE.getTrap(connection, i));
    }
    count = Client.INSTANCE.getThiefCount(connection);
    thiefs = new Thief[count];
    for(int i = 0; i < count; i++)
    {
      thiefs[i] = new Thief(Client.INSTANCE.getThief(connection, i));
    }
    count = Client.INSTANCE.getThiefTypeCount(connection);
    thiefTypes = new ThiefType[count];
    for(int i = 0; i < count; i++)
    {
      thiefTypes[i] = new ThiefType(Client.INSTANCE.getThiefType(connection, i));
    }
    count = Client.INSTANCE.getTrapTypeCount(connection);
    trapTypes = new TrapType[count];
    for(int i = 0; i < count; i++)
    {
      trapTypes[i] = new TrapType(Client.INSTANCE.getTrapType(connection, i));
    }

    if(!initialized)
    {
      initialized = true;
      init();
    }
    return run();
  }


  ///The width of the total map.
  int mapWidth()
  {
    return Client.INSTANCE.getMapWidth(connection);
  }
  ///The height of the total map.
  int mapHeight()
  {
    return Client.INSTANCE.getMapHeight(connection);
  }
  ///The current turn number.
  int turnNumber()
  {
    return Client.INSTANCE.getTurnNumber(connection);
  }
  ///The current turn number for this round.
  int roundTurnNumber()
  {
    return Client.INSTANCE.getRoundTurnNumber(connection);
  }
  ///The maximum number of Thieves allowed per player.
  int maxThieves()
  {
    return Client.INSTANCE.getMaxThieves(connection);
  }
  ///The maximum number of Traps allowed per player.
  int maxTraps()
  {
    return Client.INSTANCE.getMaxTraps(connection);
  }
  ///The id of the current player.
  int playerID()
  {
    return Client.INSTANCE.getPlayerID(connection);
  }
  ///What number game this is for the server.
  int gameNumber()
  {
    return Client.INSTANCE.getGameNumber(connection);
  }
  ///What round of the game this is.
  int roundNumber()
  {
    return Client.INSTANCE.getRoundNumber(connection);
  }
  ///The scarabs given to a player to purchase traps per round.
  int scarabsForTraps()
  {
    return Client.INSTANCE.getScarabsForTraps(connection);
  }
  ///The scarabs given to a player to purchase thieves per round.
  int scarabsForThieves()
  {
    return Client.INSTANCE.getScarabsForThieves(connection);
  }
  ///The number of won rounds required to win.
  int roundsToWin()
  {
    return Client.INSTANCE.getRoundsToWin(connection);
  }
  ///The maximum number of round turns before a winner is decided.
  int roundTurnLimit()
  {
    return Client.INSTANCE.getRoundTurnLimit(connection);
  }
  ///The number of sarcophagi each player will start with each round.
  int numberOfSarcophagi()
  {
    return Client.INSTANCE.getNumberOfSarcophagi(connection);
  }
}
